src/eval.rs

```diff
@@ -279,6 +282,20 @@ pub fn create_ecx<'mir, 'tcx: 'mir>(
     Ok((ecx, ret_place))
 }
 
+// This is potentially a performance hazard.
+// Factoring it into its own function lets us keep an eye on how much it shows up in a profile.
+fn set_current_span<'mir, 'tcx: 'mir>(ecx: &mut MiriEvalContext<'mir, 'tcx>) {
+    let current_span = Machine::stack(&ecx)
+        .into_iter()
+        .rev()
+        .find(|frame| ecx.machine.is_local(frame.instance.def_id()))
+        .map(|frame| frame.current_span())
+        .unwrap_or(rustc_span::DUMMY_SP);
+    if let Some(sb) = ecx.memory.extra.stacked_borrows.as_mut() {
+        sb.get_mut().current_span = current_span;
```

Comment:
I wonder if there is some reasonable way we could just have the span available where it is currently missing? I assume this is for `memory_read` and `memory_written` and it seems not unreasonable to give them a `TyCtxtAt` which contains a span.

Reply:
Not entirely sure what you're getting at, but I have so many feelings about this code.

I've generally tried to avoid modifying rustc to provide the functionality I want here. But we definitely need some more state somewhere, I feel like I'm starting to shove state where it doesn't really belong.

This code produces `DUMMY_SP` when there isn't a _local_ span available. I'm not sure what span to select when there is nothing local. The topmost span? And this code is a bit constrained because it needs the array of local crates available to deduce which span it should be considering.

Reply:
There is `ecx.cur_span()` to get the best approximation of the "current" span.

However, I did not realize that you are deliberately computing another span than that here. The problem is that inside the read/write handlers we cannot have access to the entire machine state since there are outstanding borrows on particular allocations. We *could* have access to the stack though... that might not even require changes to rustc, if we can move the `threads` field from `Evaluator` to `MemoryExtra`. (There is not really a conceptual distinction between those two, we just put things into `MemoryExtra` whenever that is required.)

Reply:
Here is how I think you can entirely avoid this "set_cur_span" thing: https://github.com/RalfJung/miri/commit/a20fb716cff3d946ab93fb840bea0284601d62a2

Reply:
I need to thread a bit more logic through to have access to the local crate list, but the logic is factored better!